### ISSUE TITLE
Creating a policy with policyName="" is possible, but can't be referred to by the "trusted-types" CSP directive

### ISSUE THREAD

**Issue by mbrodesser-Igalia** [w3c/trusted-types#466]
E.g. https://jsfiddle.net/q5kmL492/ is possible.

https://w3c.github.io/trusted-types/dist/spec/#trusted-types-csp-directive requires the policy-name to consist of at least one character.

That might be annoying when one writes multiple policies named `""` and wants to limit trusted-types to those policies later.

**Comment by mbrodesser-Igalia** [comment]
Adding a keyword `'allow-unnamed'` would fix this.

**Comment by lukewarlow** [comment]
This feels like it shouldn't be allowed? But if we reject unamed policies that might be a compat risk?

**Comment by mbrodesser-Igalia** [comment]
> This feels like it shouldn't be allowed? But if we reject unamed policies that might be a compat risk?

There are use-cases where policy-names are irrelevant. E.g. when allowing all policies via the wildcard `trusted-types *` (https://w3c.github.io/trusted-types/dist/spec/#trusted-types-csp-directive).

**Comment by bkardell** [comment]
I would like to understand if people really do this... Who might have some experience with how common/good an idea (or even just 'why') people would do an unnamed policy? @koto ?

**Comment by otherdaniel** [comment]
> Ww always used a policy name, but they are indeed optional (and only relevant if one guards policy creation by name with `trusted-types` directive).
> 
> @otherdaniel, can we add a use counter for unnamed policies?

Done. (TrustedTypesCreatePolicyWithEmptyName; not sure yet which release it'll appear in.)

**Comment by koto** [comment]
https://chromestatus.com/metrics/feature/timeline/popularity/4897 shows  results in the range of `0.000001` page loads. Just checking with you, @otherdaniel that it's a threshold low enough that we could remove the support for empty policy name?

**Comment by koto** [comment]
Tentatively created https://github.com/w3c/trusted-types/pull/560.

Note that it's still possible to create policies that can not be referred to by CSP, as CSP syntax limits us to https://w3c.github.io/trusted-types/dist/spec/#tt-policy-name. Disallowing creating such policies likely has much bigger backwards compatibility risk though.

**Comment by fred-wang** [comment]
I believe we still don't have tests for that case too.

**Comment by lukewarlow** [comment]
> Note that it's still possible to create policies that can not be referred to by CSP, as CSP syntax limits us to https://w3c.github.io/trusted-types/dist/spec/#tt-policy-name. Disallowing creating such policies likely has much bigger backwards compatibility risk though.

I think that's fine fwiw. No real benefit to limiting to just the csp valid names (compared to the compat effort etc).

I'm not fully sold that requiring a non-empty string is super important either but it's simple enough and presumably compat is okay?

**Comment by lukewarlow** [comment]
Seems to peak around 0.00015% is that acceptable for chromes compat purposes.

**Comment by fred-wang** [comment]
Please correct me if I'm wrong, but I understand:

- There is a backward risk for forbidding empty name.
- An empty name can make sense in cases when we don't actually care about having a name (i.e. we use a wildcard or no `trusted-types` directive), for example for quick testing.
- This can feel weird and could to lead to unexpected failures if we don't allow-duplicates, so it could make sense not to allow that.

I wonder what's current position from folks at Google, Apple and Mozilla?

**Comment by annevk** [comment]
At this point we should probably just go with the implementations? It doesn't seem like there's a strong case either way.

**Comment by lukewarlow** [comment]
I agree. Could even add a warning to devtools if your policy name isn't csp compatible and then at least it's somewhat communicated.
